iothub/device/src/Pipeline/ClientTransportStateMachine.cs

```diff
@@ -0,0 +1,96 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.Azure.Devices.Client.Transport
+{
+    internal enum ClientTransportState
+    {
+        Closed,
+        Opening,
+        Open,
+        Closing,
+    }
+
+    internal enum ClientStateAction
+    {
+        OpenStart,
+        OpenSuccess,
+        OpenFailure,
+        CloseStart,
+        CloseComplete,
+        ConnectionLost,
+    }
+
+    internal class ClientTransportStateMachine
+    {
+        internal class StateTransition
+        {
+            private readonly ClientTransportState _currentState;
+            private readonly ClientStateAction _nextAction;
+
+            internal StateTransition(ClientTransportState state, ClientStateAction action)
+            {
+                _currentState = state;
+                _nextAction = action;
+            }
+
+            public override bool Equals(object obj)
+            {
+                return obj is StateTransition other
+                    && other._currentState == _currentState && other._nextAction == _nextAction;
+            }
+
+            public override int GetHashCode()
+            {
+                return _currentState.GetHashCode() + _nextAction.GetHashCode();
+            }
+        }
+
+        private readonly Dictionary<StateTransition, ClientTransportState> _transitions;
+        private readonly object _stateTransitionLock = new();
+        private ClientTransportState _currentState;
+
+        internal ClientTransportStateMachine()
+        {
+            _currentState = ClientTransportState.Closed;
+            _transitions = new Dictionary<StateTransition, ClientTransportState>
```

Comment:
I think this could be static. It doesn't change. It is a set of facts. Right?

Reply:
sure